native/java/org/jpype/JPypeContext.java

```diff
@@ -100,17 +104,131 @@ public static JPypeContext createContext(long context, ClassLoader bootLoader)
     instance.typeManager.init();
     instance.referenceQueue.start();
 
+    // Install a shutdown hook to clean up Python resources.
+    Runtime.getRuntime().addShutdownHook(new Thread(new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        instance.watchShutdown();
+      }
+    }));
+
     return instance;
   }
 
+  /**
+   * Shutdown and remove all Python resources.
+   *
+   * This hook is only called after the last user thread has died. Thus the only
+   * remaining connections are proxies that were attached to the JVM shutdown
+   * hook, the reference queue, and the typemanager.
+   *
+   * This routine will try to take out the last connections in an orderly
+   * fashion. Inherently this is a very dangerous time as portions of Java have
+   * already been deactivated.
+   */
+  private void watchShutdown()
+  {
+    // Try to yield in case there is a race condition.  The user
+    // may have installed a shutdown hook, but we cannot verify
+    // the order that shutdown hook threads are executed.  Thus we will
+    // try to intentionally lose the race.
```

Comment:
lose the race by waiting (n**2 + n)/2 = 15 milli-seconds?

Reply:
Its late already, I replaced the 1 with an i :) Where the heck are my glasses=)

Reply:
I know that it seems odd to wait 1 ms but when the threading scheduler sees a sleep or a yield it should check to see if there are any other threads ready before coming back to this thread.    All of the threads are born during the shutdown hook and we just need to stall long enough for any unaccounted threads to hit the atomic integer increment.   Once the first one does we are safe to proceed.  

I debated pulling a longer sleep.  But most cases don't need this and it is just adding time to our shutdown.  I could also have implemented a find all Java threads code here.   But I didn't want to risk an exception should module protections be tripped.   (Yes there are a lot of horrible things to worry about here.)